language/move-lang/src/parser/syntax.rs

```diff
@@ -687,6 +743,29 @@ fn parse_call_args<'input>(tokens: &mut Lexer<'input>) -> Result<Spanned<Vec<Exp
 fn parse_exp<'input>(tokens: &mut Lexer<'input>) -> Result<Exp, Error> {
     let start_loc = tokens.start_loc();
     let exp = match tokens.peek() {
+        Tok::Fun => {
+            // Parse lambda, which is only valid in specification expressions.
+            // We currently put a lambda into a block expression of the form
+            // `Block([Declare(..)], Exp)`. The specification language
+            // accepts regular blocks only if they have bindings, so the lambda usage
+            // of blocks is distinguishable.
+            require_spec_context(tokens, |tokens| {
+                (
+                    current_token_loc(tokens),
+                    "`fun (..) <exp>`) only allowed in specifications".to_string(),
```

Comment:
This error should be done in expansion, not in parsing. In other words. The parser should not need to have a mode. Any "mode" behavior should be done in expansion

Reply:
Alright, rewrote the PR (quite significantly) to satisfy this request.